### PR TITLE
convert paths to arrays

### LINKED ISSUE #6: Paths as arrays
I think it would be a good idea to convert our notion of paths to arrays. Currently, we implement an object containing a property and index, however, it would be a good idea to convert our codebase to using arrays, as this would support a more general use case of arbitrary long paths. For example, we might want to implement a general path function, which finds a path of arbitrary length, in which case, we would need an array.

### PR DESCRIPTION
**Pull request by adamjanicki2**
Converts all paths to arrays and closes #6

### CODE REVIEW

**Review by LeaVerou — CHANGES_REQUESTED** [review]
It's not that simple I'm afraid 😄 

If we’re using paths, we can't be doing things that assume the path is of length <=2 like: 

```
const [property, index] = path;
// ...
parent[property][index] = newChild;
```

We need to instead implement util functions that take an object and a path and get/set it.

**Comment by LeaVerou** [comment]
Just pushed two util methods to help you do this!

**Comment by adamjanicki2** [comment]
> Just pushed two util methods to help you do this!
> 
> (Have not tested them, feel free to make any changes needed)

Just thinking here -- do we want to export the `get` helper to authors? I think it would be useful. We could stick it in `path.js` along with a function that finds the path between two nodes. The `set` helper should remain private since it becomes complicated with parent pointers and pretty much only works when we use it internally within a function like this

**Comment by LeaVerou** [comment]
> > Just pushed two util methods to help you do this!
> > (Have not tested them, feel free to make any changes needed)
> 
> Just thinking here -- do we want to export the `get` helper to authors? I think it would be useful. We could stick it in `path.js` along with a function that finds the path between two nodes. The `set` helper should remain private since it becomes complicated with parent pointers and pretty much only works when we use it internally within a function like this

Not sure what you mean, these are already exported and they can import them?

File: src/childPaths.js [hunk 1]
```diff
@@ -26,11 +26,11 @@ export default function childPaths (node) {
 			const child = node[property];
 			// When the node is an array, we want to include the index in the result
 			if (Array.isArray(child)) {
-				let childPaths = child.map((c, index) => ({node: c, property, index}));
+				let childPaths = child.map((c, index) => ({node: c, path: [property, index]}));
```
Comment (LeaVerou) [inline src/childPaths.js:29]: Is this still assuming a (max) two-part path?

**Review by LeaVerou — CHANGES_REQUESTED** [review]
I think a lot of this still assumes a shallow path.
Try to use your code with cases where to get from each node to each children you have to follow something like `foo.bar.baz`. I haven't looked *super* closely, but from a first skim, I don't think it would work.

**Comment by adamjanicki2** [comment]
> I think a lot of this still assumes a shallow path. Try to use your code with cases where to get from each node to each children you have to follow something like `foo.bar.baz`. I haven't looked _super_ closely, but from a first skim, I don't think it would work.

Would this even be allowed? In what cases could you follow two object properties and get to a child? In the case of `foo.bar.baz`, how could `baz` be a direct child of `foo`?

**Comment by adamjanicki2** [comment]
My thinking was we sort of have to enforce the fact that children can only be accessed via a property, or a property and an index in the case of an array; otherwise having arbitrary ways of accessing children would defeat the difference between children and descendants in general

**Comment by LeaVerou** [comment]
What’s the point of making the path an array then?

**Review by LeaVerou — APPROVED** [review]
Rethinking this, I think it's fine to move towards a `path` now, and make the code more generic as use cases emerge. As discussed, we should not assume that the second property is an index, since we already have use cases where it isn’t.

**Pull request merged**